umibukela/analysis.py

```diff
@@ -0,0 +1,295 @@
+
+
+# These are used for grouping. Trying to count them and group by them
+# at the same time doesn't work.
+SKIP_QUESTIONS = [['_uuid'], ['demographics_group', 'gender']]
+SKIP_GROUP_NAMES = ['meta', 'formhub']
+GENDER_COLUMN = 'demographics_group/gender'
+
+log = getLogger(__name__)
+
+
+class Element(object):
+    def __init__(self, element, path):
+        self.label = element['label']
+        self.name = element['name']
+        self.path = path + [element['name']]
+        self.pathstr = pathstr(self.path)
+
+    def __str__(self):
+        return "%s %s" % (self.__class__, self.pathstr)
+
+
+class MultipleChoice(Element):
+    def __init__(self, question, path, group_labels):
+        super(MultipleChoice, self).__init__(question, path)
+        self.options = [Option(o, self.path) for o in question['children']]
+        self.group_labels = group_labels
+
+
+class Option(Element):
+    pass
+
+
+class SelectOne(MultipleChoice):
+    pass
+
+
+class SelectAllThatApply(MultipleChoice):
+    pass
+
+
+def count_submissions(submissions):
+    results = {}
+    # per-gender counts
+    cols = ['_uuid', GENDER_COLUMN]
+    question_table = submissions.loc[:, cols]
+    gender_counts = question_table.groupby(
+        [GENDER_COLUMN]
+    ).count()
+    results['female'] = int(gender_counts.loc['female'])
+    results['male'] = int(gender_counts.loc['male'])
+    results['total'] = int(submissions.loc[:, ['_uuid']].count())
+
+    return results
+
+
+def count_options(submissions, children, path=None, group_labels=None, results=None):
+    """
+    returns nested dicts where the keys are the names of the XForm element
+    branches to each question and each option of a question. Only multiple
+    choice questions are supported.
+    """
+    path = path or []  # list of names in structure leading to current element
+    group_labels = group_labels or []  # list of labels in groups in path
+    results = results or {}  # results nested dict under construction
+    for child in children:
+        deeper_path = path + [child['name']]
+        if deeper_path in SKIP_QUESTIONS:
+            pass
+        elif child.get('type') == 'group' and child['name'] in SKIP_GROUP_NAMES:
+            pass
+        elif child.get('type') == 'group':
+            deeper_group_labels = group_labels + [child['label']]
+            results = count_options(submissions, child['children'],
+                                    deeper_path, deeper_group_labels, results)
+        elif child.get('type') == 'select one':
+            question = SelectOne(child, path, group_labels)
+            results = count_select_one(submissions, question, results)
+        elif child.get('type') == 'select all that apply':
+            question = SelectAllThatApply(child, path, group_labels)
+            results = count_select_all_that_apply(submissions, question, results)
+        else:
+            pass
+    return results
+
+
+def count_select_one(submissions, q, results):
```

Comment:
This could potentially be part of one of the classes above? Not critical, just a suggestion.